design-documents/data-format/json.md

```diff
@@ -0,0 +1,245 @@
+        }
+        return $result;
+    }
+}
+```
+
+## Problem overview
+JSON as data format can be used in other different from serialization contexts so `Magento\Framework\Serialize\Serializer\Json` class implementation is not sufficient enough to cover all those scenarios, e.g. it is impossible to pass `$options` or `$depth`.
+
+Native PHP [json_encode](https://www.php.net/manual/en/function.json-encode.php) function:
+``` php
+string json_encode ( mixed $value [, int $options = 0 [, int $depth = 512 ]] )
+```
+
+Native PHP [json_decode](https://www.php.net/manual/en/function.json-decode.php) function:
+``` php
+mixed json_decode ( string $json [, bool $assoc = FALSE [, int $depth = 512 [, int $options = 0 ]]] )
+```
+
+On the other hand native PHP implementation requires us to call `json_last_error` in order to handle errors instead of throwing an exception. It was [changed in PHP 7.3](https://wiki.php.net/rfc/json_throw_on_error), but this is still not the default behaviour.
+
+## Solution
+
+Since we want to protect our platform from any possible security vulnerabilities related to the code we do not control and have expected implementation we propose to add new dedicated contract for JSON.
+
+``` php
+interface JsonEncoderInterface
+{
+    /**
+     * Encode data into JSON string
+     *
+     * @param string|int|float|bool|array|null $data
+     * @param int $option
+     * @param int $depth
+     * @return string
+     * @throws \InvalidArgumentException
+     */
+    public function encode($data, int $option, int $depth);
+
+    /**
+     * Decode the given JSON string
+     *
+     * @param string $string
+     * @param bool $assoc
+     * @param int $option
+     * @param int $depth
+     * @return string|int|float|bool|array|null
+     * @throws \InvalidArgumentException
+     */
+    public function decode($string, bool $assoc, int $option, int $depth);
+
+    //it's possible to add some other methods
+}
+```
+
+Implementation would use low-level PHP `json_encode`/`json_decode` under the hood and will throw `InvalidArgumentException` if encoding/decoding cannot be performed.
+New interface introduction means that we need to refactor all the existing business logic which currently depends on JSON class, and substitute it with this new service `JsonEncoderInterface`, because having two contracts means that it's incorrect to depend directly on JSON object (particular implementation) for JSON encoding purposes, that's a violation of polymorphism.
+
+#### Pros
+- Correct from OOP point of view as we have two independent interfaces for two different business operations.
+- Follows the Interface Segregation Principle (SOLID).
+- Get us closer to the desired state of serializer/encoding design.
+- Proper error handling.
+
+#### Cons
+- Backward incompatible as introduces a lot of refactoring substituting existing direct usages of JSON object with newly introduced contract `JsonEncoderInterface`.
+
+### Other possible solutions
```

Comment:
Discussed with @Vinai and @buskamuza and narrowed down to 2 options.

Deprecated common interface in the patch release. Remove common interface in minor release and add optional parameter to Json class.

Deprecate common interface. Do not remove interface in the minor release, keep it for some time. Introduce new methods SerializeWithOptions/UnserializeWithOptions in Json class.

Could you please update proposal with these options, so we can discuss them on the next architecture meeting?

Reply:
Discussed on ad hoc meeting with @paliarush and @joni-jones and decided to deprecate interface in the next minor release, provide new JsonSerializer with serialize/unserialize methods and in minor release after next remove the interface.

Reply:
Will the deprecated interface stay around, or are you still planing to remove it later? If so, when?
What is the benefit of providing a new `JsonSerializer` class over adding two methods to the existing `Json` class? Will the `Json` class be deprecated, too?

From my perspective using the existing `Json` class is preferable, instead of introducing a new class.
To allow specifying serialization options, two new dedicated methods should be added. This keeps backward compatibility and provides a clean interface.

The biggest benefit is that all the existing code that uses the `Json` class does not need to be refactored. This kind of refactoring is useless busy work that doesn't add any value in return.
Another benefit is that it's less confusing to have a single implementation rather then two, even if one of them is deprecated.

Personally I find a class with the methods `Json::serialize` and `Json::serializeWithOptions`  (plus matching deserialization methods) far more expressive and clean compared to a class `JsonSerializer::serialize` with optional options arguments.

Anyway, just my .02$. I really feel I'm repeating myself.
In a nutshell, I advocate avoiding useless bc breaks (to avoid angering developers building on the platform), and simple is better.